src/rdma.c

```diff
@@ -1689,25 +1697,24 @@ static int rdmaProcessPendingData(void) {
     listNode *ln;
     rdma_connection *rdma_conn;
     connection *conn;
-    listNode *node;
     int processed;
 
     processed = listLength(pending_list);
     listRewind(pending_list, &li);
     while ((ln = listNext(&li))) {
         rdma_conn = listNodeValue(ln);
         conn = &rdma_conn->c;
-        node = rdma_conn->pending_list_node;
 
         /* a connection can be disconnected by remote peer, CM event mark state as CONN_STATE_CLOSED, kick connection
          * read/write handler to close connection */
         if (conn->state == CONN_STATE_ERROR || conn->state == CONN_STATE_CLOSED) {
-            listDelNode(pending_list, node);
-            /* do NOT call callHandler(conn, conn->read_handler) here, conn is freed in handler! */
-            if (conn->read_handler) {
-                conn->read_handler(conn);
-            } else if (conn->write_handler) {
-                conn->write_handler(conn);
+            listDelNode(pending_list, rdma_conn->pending_list_node);
+            rdma_conn->pending_list_node = NULL;
+            /* in case read_handler is NULL, we have a chance to invoke write_handler.
+               if the connection is closed successfully in read_handler, the returned 0
+               will not invoke write_handler any more. */
```

Comment:
The comment doesn't seem aligned with the logic. Whether we invoke `write_handler` is not predicated on `read_hander` being NULL or not but whether invoking `read_handler` causes the connection to be closed. How about the following?

```suggestion
            /* Invoke both read_handler and write_handler, unless read_handler 
               returns 0, indicating the connection has closed, in which case 
               write_handler will be skipped. */
```